### PR TITLE
gnrc: simplify hdr_build functions

### PR DESCRIPTION
**Pull request by OlegHahm**
Directly using 16 bit port numbers instead of casting uint8_t pointers.

### CODE REVIEW

**Comment by kaspar030** [comment]
Wow, someone was planning ahead! ;) +1

**Comment by cgundogan** [comment]
Just wondering, and maybe unrelated to this PR: do we want to support 8-bit port numbers in the future?

**Comment by authmillenon** [comment]
Would gladly appreciate if this would be done for `ipv6_addr_t` ;-).

File: sys/net/gnrc/netreg/gnrc_netreg.c [hunk 1]
```diff
@@ -148,12 +148,20 @@ gnrc_pktsnip_t *gnrc_netreg_hdr_build(gnrc_nettype_t type, gnrc_pktsnip_t *paylo
...
+            }
 #endif
 #ifdef MODULE_GNRC_UDP
 
         case GNRC_NETTYPE_UDP:
-            return gnrc_udp_hdr_build(payload, src, src_len, dst, dst_len);
+            {
+            uint16_t src_port = *((uint16_t *)src);
+            uint16_t dst_port = *((uint16_t *)dst);
+            return gnrc_udp_hdr_build(payload, src_port, dst_port);
```
Comment (authmillenon) [inline sys/net/gnrc/netreg/gnrc_netreg.c]: Error check from https://github.com/RIOT-OS/RIOT/pull/5118/files#diff-a743a0dcf90d0eb32c837f1a047918c1L283 needs to be moved here now.

**Comment by OlegHahm** [comment]
> Just wondering, and maybe unrelated to this PR: do we want to support 8-bit port numbers in the future?

What do you mean by 8-bit port numbers?

**Comment by OlegHahm** [comment]
> Would gladly appreciate if this would be done for ipv6_addr_t gnrc_ipv6_hdr_build(), too ;-).

Will do.

**Comment by OlegHahm** [comment]
updated

File: sys/include/net/gnrc/ipv6/hdr.h [hunk 2]
```diff
@@ -37,19 +37,14 @@ extern "C" {
...
- *                      must be `sizeof(ipv6_addr_t)`.
  *
  * @return  The an IPv6 header in packet buffer on success.
  * @return  NULL on error.
  */
-gnrc_pktsnip_t *gnrc_ipv6_hdr_build(gnrc_pktsnip_t *payload,
-                                    uint8_t *src, uint8_t src_len,
-                                    uint8_t *dst, uint8_t dst_len);
+gnrc_pktsnip_t *gnrc_ipv6_hdr_build(gnrc_pktsnip_t *payload, uint8_t *src,
+                                    uint8_t *dst);
```
Comment (authmillenon) [inline sys/include/net/gnrc/ipv6/hdr.h]: Why not use `ipv6_addr_t` pointers?

Comment (OlegHahm) [inline sys/include/net/gnrc/ipv6/hdr.h]: true, will update

**Comment by cgundogan** [comment]
>What do you mean by 8-bit port numbers?

I mean that 16-bit port numbers may be too much for certain IoT use cases. Do we plan to make that configurable (by compile time e.g.) to reduce that to 8-bit instead somewhen in the future?

**Comment by kaspar030** [comment]
> What do you mean by 8-bit port numbers?
> 
> I mean that 16-bit port numbers may be too much for certain IoT use cases. Do we plan to make that configurable (by compile time e.g.) to reduce that to 8-bit instead somewhen in the future?

I don't think we should make that part of the protocol compile-time configurable...
Doesn't 6lowpan define a set of "short" UDP ports, but they still start at a value higher than 256?

**Comment by OlegHahm** [comment]
NHC allows to compress a certain range of UDP ports. Compressing them internally seems to be overkill to me.

**Comment by cgundogan** [comment]
IPv6 UDP NHC can compress the ports to 4-bits each if they are in the range of `0xf0b0` to `0xf0bf`. But currently all our APIs will use 16-bits for each port (src,dst). I am just wondering if it might make sense to make our (future) APIs configurable for that

**Comment by OlegHahm** [comment]
I don't think it makes sense to use compressed representation in the API.

**Comment by authmillenon** [comment]
@cgundogan that's out-of-scope for IPv6 and purely a 6LoWPAN thing. I would keep it there and not put 6LoWPAN specific stuff in our UDP implementation

**Comment by OlegHahm** [comment]
Addressed @authmillenon's comment.

**Comment by OlegHahm** [comment]
Thanks for the review - I think this PR now simplifies the code a bit.

**Comment by authmillenon** [comment]
Please squash (and change title of PR appropriately ;-))

**Comment by OlegHahm** [comment]
done

File: sys/net/gnrc/netreg/gnrc_netreg.c [hunk 3]
```diff
@@ -143,17 +144,31 @@ gnrc_pktsnip_t *gnrc_netreg_hdr_build(gnrc_nettype_t type, gnrc_pktsnip_t *paylo
 #ifdef MODULE_GNRC_IPV6
 
         case GNRC_NETTYPE_IPV6:
-            return gnrc_ipv6_hdr_build(payload, src, src_len, dst, dst_len);
+            assert(src_len == sizeof(ipv6_addr_t));
+            assert(dst_len == sizeof(ipv6_addr_t));
```
Comment (authmillenon) [inline sys/net/gnrc/netreg/gnrc_netreg.c]: Please document these preconditions in the doxygen.

**Comment by authmillenon** [comment]
One (nitpicky) remark. Otherwise I give my ACK when Murdock is happy.

**Comment by authmillenon** [comment]
(you may squash the doc fix immediately).

File: sys/net/gnrc/netreg/gnrc_netreg.c (hunk 3 above)
Comment (OlegHahm) [inline sys/net/gnrc/netreg/gnrc_netreg.c]: Hm, wondering about how to document this, I begin to wonder if we need this function at all. What is the purpose of this function?

**Comment by OlegHahm** [comment]
CI was failing because of missing const specifiers. Unsure if I should squash this last commit.

**Comment by OlegHahm** [comment]
I couldn't think of a reason where this generic header build function would be needed, hence I removed it.

**Comment by authmillenon** [comment]
I think the original idea was to be layer agnostic, but that proved to be impossible at least for TCP/IP.  Any objections @haukepetersen?

**Comment by authmillenon** [comment]
Please rebase.

**Comment by OlegHahm** [comment]
rebased to master

**Comment by authmillenon** [comment]
ACK and go.

**Comment by authmillenon** [comment]
(when Murdock is happy).

**Comment by OlegHahm** [comment]
Fixed a typo in zep that Murdock reported and squashed immediately.

**Comment by OlegHahm** [comment]
args, fixed the wrong parameter - another try.

**Pull request merged**